File: 984-most-stones-removed-with-same-row-or-column/most-stones-removed-with-same-row-or-column.py
class DSU:
    def __init__(self,n):
        self.parent=[i for i in range(n)]
        self.rank=[0]*n
    def find(self,x):
        if self.parent[x]==x:
            return self.parent[x]
        self.parent[x]=self.find(self.parent[x])
        return self.parent[x]
    def union(self,a,b):
        py=self.find(a)
        pu=self.find(b)
        if py==pu:
            return 
        if self.rank[py]<self.rank[pu]:
            self.parent[py]=pu
        elif self.rank[py]>self.rank[pu]:
            self.parent[pu]=py
        else:
            self.parent[py]=pu
            self.rank[pu]+=1


class Solution:
    def removeStones(self, stones: List[List[int]]) -> int:
        n=len(stones)
        dsu=DSU(n)

        for i in range(n):
            for j in range(i+1,n):
                if stones[i][0]==stones[j][0] or stones[i][1]==stones[j][1]:
                    dsu.union(i,j)


        connected=set()
        for i in range(n):
            connected.add(dsu.find(i))
        return n-len(connected)

File: 984-most-stones-removed-with-same-row-or-column/most-stones-removed-with-same-row-or-column.cpp
class DSU {
public:
    vector<int> parent;
    vector<int> rank;

    DSU(int n) {
        parent.resize(n);
        rank.resize(n, 0);

        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }

        return parent[x];
    }

    void unite(int a, int b) {
        int pa = find(a);
        int pb = find(b);

        if (pa == pb)
            return;

        if (rank[pa] < rank[pb]) {
            parent[pa] = pb;
        }
        else if (rank[pa] > rank[pb]) {
            parent[pb] = pa;
        }
        else {
            parent[pb] = pa;
            rank[pa]++;
        }
    }
};


class Solution {
public:
    int removeStones(vector<vector<int>>& stones) {

        int n = stones.size();

        DSU dsu(n);

        // Connect stones having same row or column
        for (int i = 0; i < n; i++) {

            for (int j = i + 1; j < n; j++) {

                if (stones[i][0] == stones[j][0] ||
                    stones[i][1] == stones[j][1]) {

                    dsu.unite(i, j);
                }
            }
        }

        // Count connected components
        int components = 0;

        for (int i = 0; i < n; i++) {
            if (dsu.find(i) == i) {
                components++;
            }
        }

        return n - components;
    }
};